cli/azd/pkg/infra/provisioning/terraform/terraform_provider.go

```diff
@@ -0,0 +1,595 @@
+				ParameterFilePath: t.parametersFilePath(),
+				PlanFilePath:      t.planFilePath(),
+			}
+			if !isRemoteBackendConfig {
+				deploymentDetails.localStateFilePath = t.localStateFilePath()
+			}
+
+			result := DeploymentPlan{
+				Deployment: *deployment,
+				Details:    deploymentDetails,
+			}
+
+			asyncContext.SetResult(&result)
+		})
+}
+
+// Deploy the infrastructure within the specified template through terraform apply
+func (t *TerraformProvider) Deploy(ctx context.Context, deployment *DeploymentPlan, scope infra.Scope) *async.InteractiveTaskWithProgress[*DeployResult, *DeployProgress] {
+	return async.RunInteractiveTaskWithProgress(
+		func(asyncContext *async.InteractiveTaskContextWithProgress[*DeployResult, *DeployProgress]) {
+			t.console.Message(ctx, "Locating plan file...")
+			currentSubscription, err := t.ensureEnvSubscription(ctx)
+			if err != nil {
+				asyncContext.SetError(fmt.Errorf("failed to set az subscription , err: %w", err))
+				return
+			}
+			if currentSubscription != "" {
+				defer t.setAZSubscription(ctx, currentSubscription)
+			}
+
+			modulePath := t.modulePath()
+			terraformDeploymentData := deployment.Details.(TerraformDeploymentDetails)
+			isRemoteBackendConfig, err := t.isRemoteBackendConfig()
+			if err != nil {
+				asyncContext.SetError(fmt.Errorf("reading backend config: %w", err))
+				return
+			}
+
+			t.console.Message(ctx, "Deploying terraform template...")
+			err = asyncContext.Interact(func() error {
+				applyArgs, err := t.createApplyArgs(isRemoteBackendConfig, terraformDeploymentData)
+				if err != nil {
+					return err
+				}
+
+				runResult, err := t.cli.Apply(ctx, modulePath, applyArgs...)
+				if err != nil {
+					return fmt.Errorf("template Deploy failed: %s , err:%w", runResult, err)
+				}
+
+				return nil
+			})
+
+			if err != nil {
+				asyncContext.SetError(err)
+				return
+			}
+
+			// Set the deployment result
+			outputs, err := t.createOutputParameters(ctx, modulePath, isRemoteBackendConfig)
+			if err != nil {
+				asyncContext.SetError(fmt.Errorf("create terraform template failed: %w", err))
+				return
+			}
+
+			currentDeployment := deployment.Deployment
+			currentDeployment.Outputs = outputs
+			result := &DeployResult{
+				Deployment: &currentDeployment,
+			}
+
+			asyncContext.SetResult(result)
+		})
+}
+
+// Destroys the specified deployment through terraform destroy
+func (t *TerraformProvider) Destroy(ctx context.Context, deployment *Deployment, options DestroyOptions) *async.InteractiveTaskWithProgress[*DestroyResult, *DestroyProgress] {
+	return async.RunInteractiveTaskWithProgress(
+		func(asyncContext *async.InteractiveTaskContextWithProgress[*DestroyResult, *DestroyProgress]) {
+			// discuss : check if you want to set auto-destroy . CAN WE PASS VALUES BACK AND FORTH BETWEEN THE AZD CONTEXT AND PROCESS
+			os.Setenv("TF_DATA_DIR", t.dataDirPath())
+
+			currentSubscription, err := t.ensureEnvSubscription(ctx)
```

Comment:
I am assuming that we are doing this because by default the `azurerm` provider uses the current `az` cli subscription, which may differ from what the user has configured with `azd`.  I really like the use of `defer` here to make sure we undo the operation when we are done.

I noticed that the provider documentation says it can source the ID from the environment variable `ARM_SUBSCRIPTION_ID`.  Did we consider using that instead of this approach?  It might be less error prone and quicker.

Reply:
@HadwaAbdelhalem - can you confirm this? I recall you mentioning the env vars was either all or nothing?

Reply:
great call out. updated the code to set ARM_SUBSCRIPTION_ID on the cli